Add comprehensive LLM and Copilot instruction files to CSDK (#239)
Sitecore/content-sdk: copilot-instructions.md

# GitHub Copilot Instructions for Sitecore Content SDK

## Project Overview

This is the **Sitecore Content SDK** - a comprehensive TypeScript/JavaScript SDK for building modern web applications with Sitecore XM Cloud. The project provides core functionality, React components, Next.js integration, and CLI tools for scaffolding applications.

### Tech Stack
- **Language**: TypeScript (Node LTS)

> - **Runtime**: Node LTS or higher

Reviewer: We only rely on LTS version and don't really use "latest", probably, let's update it
```suggestion
- **Runtime**: Node LTS
```

- **Build**: `tsc` -> `dist/`, templates bundled via `scripts/build-templates.ts`
- **Testing**: Mocha + Sinon + Chai, coverage via `nyc`
- **Lint/format**: ESLint + Prettier
- **Package Manager**: Yarn 3.1.0

### Core Packages
- `@sitecore-content-sdk/core` - Core SDK functionality
- `@sitecore-content-sdk/react` - React components and hooks
- `@sitecore-content-sdk/nextjs` - Next.js integration and middleware

> - `@sitecore-content-sdk/cli` - CLI tools and scaffolding
> - `@sitecore-content-sdk/create-content-sdk-app` - Scaffolding CLI & templates

Reviewer: Both CLI's are marked as "scaffolding". May be "create-content-sdk-app" can be rephrased a bit to be clear that it's used to scaffold templates?
```suggestion
- `@sitecore-content-sdk/cli` - CLI tools and scaffolding
- `@sitecore-content-sdk/create-content-sdk-app` - CLI for template scaffolding
```

## Repository Structure

```
content-sdk/
├── packages/

> │   ├── create-content-sdk-app/    # Scaffolding CLI & templates

Reviewer: Should be updated together with the changes above

│   ├── core/                      # Core SDK functionality
│   ├── nextjs/                    # Next.js-specific SDK

> │   ├── react/                     # React components

Reviewer: Does it make sense to align this title with the title for "nextjs" package?
```suggestion
│   ├── react/                     # React-specific SDK
```

│   └── cli/                       # Additional CLI tools
├── .cursor/rules/                 # AI coding guidance
└── samples/                       # Example applications
```

### Key Directories

> - **Sources**: `src/**` (within create-content-sdk-app package)

Reviewer: Hm, why do we need to mention "create-content-sdk-app" package here explicitly? I believe it's about all our packages

- **Never edit**: `dist/**` (compiled output)
- **Templates**: Copied to generated apps, self-contained, use `.env.*.example` for env values
- **Initializers**: Each exposes `init(args)`, reuse common processes/utilities

## Code Style

### Vibe-Coding Principles

Core Philosophy:
- Write clean, modular, and idiomatic code
- Prefer declarative over imperative patterns
- Make code readable and self-documenting
- TypeScript-first development approach

Code Organization:

> - Use Node LTS+ features when stable

Reviewer: ```suggestion
- Use Node LTS
```

### Performance

Optimization Patterns:
- Memoize components with React.memo when appropriate
- Lazy-load non-critical modules: `const Component = lazy(() => import('./Component'))`
- Use useCallback and useMemo for expensive operations

TypeScript:
- Enable strict mode in tsconfig.json
- Prefer type assertions over any: `value as ContentItem`
- Use discriminated unions for complex state management

### Documentation

JSDoc Comments:
- All new functions, interfaces, classes must have JSDoc style comments
- Include @param tags for all parameters with types and descriptions
- Include @returns tag for return values with type and description
- Use descriptive comments that explain the purpose and behavior
- Follow existing Content SDK JSDoc patterns

## Sitecore SDK Rules

### XM Cloud Integration

Configuration:
- Always use environment variables for API endpoints and keys
- Never hardcode API keys in source code
- Use `.env.example` files to document required environment variables
- Prefer `sitecore.config.ts` for centralized configuration

```typescript
// sitecore.config.ts
export const sitecoreConfig = {
  sitecoreApiHost: process.env.SITECORE_API_HOST || '',

>   sitecoreApiKey: process.env.SITECORE_API_KEY || '',

Reviewer: Would be better to provide a default example using client/server context id:
Also our interface is a bit different:
```ts
api: {
   edge: { ... }
}
```

Reviewer: This one is not addressed, you can see a lot of entries

>   siteName: process.env.SITECORE_SITE_NAME || 'default',

Reviewer: ```suggestion
  defaultSite: process.env.SITECORE_SITE_NAME || 'default',
```

Reviewer: Not addressed

### Component Development

Sitecore Component Naming:
- Use descriptive, feature-based names: `HeroWithContent`, `ProductListing`, `ContentBlockGrid`
- Follow PascalCase convention
- Include component type in name when helpful: `LayoutContainer`, `ContentRenderer`

Component Registration:
- Register components in the component map
- Use consistent component names across registration and files
- Include TypeScript types for all component props

```typescript
// Component props interface
interface ContentBlockProps {
  fields: {
    title: Field;
    content: Field;
    image: Field;
  };
}
```

### Content Management

Field Handling:
- Always validate field existence before rendering
- Use helper functions for common field operations
- Handle empty/null fields gracefully
- Prefer Sitecore field components over manual rendering

```typescript
// Good: Using Sitecore field components
<Text field={fields?.title} tag="h1" />
<RichText field={fields?.content} />

// Avoid: Manual field value extraction unless necessary
```

Content SDK Client Methods:
- Prefer using `scClient.getPage()` for page data and layout fetching
- Use `scClient.getDictionary()` for localized content

> - Leverage `scClient.getErrorPages()` for error page handling

Reviewer: ```suggestion
- Leverage `scClient.getErrorPage()` for error page handling
```

Reviewer: Not addressed here: https://github.dev/Sitecore/content-sdk/blob/4e93a82a0c98a20f60f1b49b67558fc3878eb79a/.cursor/rules/sitecore.mdc#L101

> - Use `scClient.getPreviewPage()` for preview mode content

Reviewer: ```suggestion
- Use `scClient.getPreview()` for preview mode content
```

Reviewer: Not addressed here: https://github.dev/Sitecore/content-sdk/blob/4e93a82a0c98a20f60f1b49b67558fc3878eb79a/.cursor/rules/sitecore.mdc#L102
